File: problem_4/problem_4.js
const find_palindrome = n => {
    return n === +n.toString().split("").reverse().join("");
}

// Clever and functional, but hard to reason with.
const find_largest = () => {
    const arr = Array(900).fill(0).map((_, i) => 999 - i);
    return arr.reduce((acc, a) => {
        acc.push(arr.filter(b => find_palindrome(a * b) ? b : null).map(n => n * a));
        return acc;
    }, []).flat().sort((a, b) => b - a)[0];
}

const find_largest_loop = () => {
    let max = 999,
        largest = 0;
    while (max > 100) {
        let max2 = max;
        if (max * max2 < largest) {
            break;
        }
        while (max2 > 100) {
            if (find_palindrome(max * max2) && (max * max2 > largest)) {
                largest = max * max2;
            }
            max2--;
        }
        max--;
    }
    return largest;
}

console.log(find_largest_loop(), "loop");
console.log(find_largest(), "reduce");

File: problem_4/problem_4.py
def find_palindrome(n):
    rev = int(''.join(map(lambda x: x ,reversed(list(str(n))))))
    return n == rev

def largest_palindrome_faster():
    max = 999
    largest_palindrome = 0
    while max > 100:
        max2 = max
        if (max * max2 < largest_palindrome):
            break
        while max2 > 100:
            if find_palindrome(max * max2) and (max * max2 > largest_palindrome):
                largest_palindrome = max * max2
            max2 -= 1
        max -= 1
    return largest_palindrome

print(largest_palindrome_faster(), "faster")

def largest_palindrome():
    max = 999
    max2 = 999
    min = 100
    palindromes = []
    while max > min:
        if (max2 < 100):
            max2 = max = max - 1
        if find_palindrome(max * max2):
            palindromes.append(max * max2)
        max2 -= 1
    return list(reversed(sorted(palindromes)))[0]

print(largest_palindrome())
# print(find_palindrome(1234))
# print(find_palindrome(8008))
